server/comms/rpc/rpctypes.go

```diff
@@ -0,0 +1,196 @@
+	// requests.
+	ID *interface{} `json:"id"`
+}
+
+// NewResponse returns a new JSON-RPC response object. Though JSON-RPC
+// specification dictates that either but not both of the result or the error
+// should be nil, NewResponse does not check this condition.
+func NewResponse(id interface{}, result interface{}, rpcErr *RPCError) (*Response, error) {
+	if !IsValidIDType(id) {
+		return nil, fmt.Errorf("the id of type '%T' is invalid", id)
+	}
+
+	marshalledResult, err := json.Marshal(result)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Response{
+		Jsonrpc: JSONRPCVersion,
+		Result:  marshalledResult,
+		Error:   rpcErr,
+		ID:      &id,
+	}, nil
+}
+
+// ID64 attempts to parse the ID into an int64. The boolean return value will
+// be false if the type ID type is incompatible.
+func (resp *Response) ID64() (int64, bool) {
+	return id64(*resp.ID)
+}
+
+// Message returns the Request wrapped in a Message.
+func (resp *Response) Message() (*Message, error) {
+	respBytes, err := json.Marshal(resp)
+	if err != nil {
+		return nil, err
+	}
+	return &Message{
+		Type:    ResponseMessage,
+		Payload: json.RawMessage(respBytes),
+	}, nil
+}
+
+// MessageType indicates the type of payload in a message.
+const (
+	InvalidMessage = iota
+	ResponseMessage
+	RequestMessage
+)
+
+// Message is the primary messaging type, and generally wraps a Response or
+// a Request as it's payload. The Message wrapper is required to accomodate
+// bi-directional requests on a single connection, i.e. the client can send
+// Requests to the server, but the server can also send Requests to the client.
+type Message struct {
+	Type    uint8           `json:"type"`
+	Payload json.RawMessage `json:"payload"`
+}
+
+// IsValidIDType checks that the ID field (which can go in any of the JSON-RPC
+// requests, responses, or notifications) is valid.JSON-RPC 2.0 only allows
+// string, number, or null, so this function restricts the allowed types to that
+// list.
+func IsValidIDType(id interface{}) bool {
+	switch id.(type) {
+	case int, int8, int16, int32, int64,
+		uint, uint8, uint16, uint32, uint64,
+		float32, float64,
+		string,
+		nil:
+		return true
+	default:
+		return false
+	}
+}
+
+// id64 tries to parse the interface{} as an int64. Only numeric types will
+// parse. If the type is non-numeric, the boolean return value will be false.
+func id64(id interface{}) (int64, bool) {
+	switch i := id.(type) {
+	case int:
+		return int64(i), true
+	case int8:
+		return int64(i), true
+	case int16:
+		return int64(i), true
+	case int32:
+		return int64(i), true
+	case int64:
+		return int64(i), true
+	case uint:
+		return int64(i), true
+	case uint8:
+		return int64(i), true
+	case uint16:
+		return int64(i), true
+	case uint32:
+		return int64(i), true
+	case uint64:
+		return int64(i), true
+	case float32:
+		return int64(i), true
+	case float64:
+		return int64(i), true
+	}
+	return -1, false
```

Comment:
Any reason why this isn't being simplified to:

```
switch id.(type){
	case int, int8, int16, int32, int64, uint,
		uint8, uint16, uint32, uint64, float32, float64:
		return int64(i), true
	default:
		return -1, false 
}
```
as was done for `IsValidIDType` ?

Reply:
https://play.golang.org/p/59bdOi2Sl7v

Reply:
Noticed why my suggestion wouldn't work, thanks. I think support for string ids should be dropped, ids could be `*uint64` which allows for nil ids for notifications and simplifies parsing. This would be supporting a subset of the json 2.0 spec with regards to ids but that works for what comms between dex client and server. Are there any plans to use string ids?

Reply:
I agree in spirit, but I'd like to stay on JSON-RPC spec as much as possible. If the client wants to use a string ID, we can handle that.

Note that for client-originating requests, `id64` is never used. The ID type is validated, but beyond that, we don't care. `id64` is only used for DEX-originating requests and the associated response.

Reply:
The spec expects an int id https://github.com/decred/dcrdex/tree/master/spec#json-rpc

Reply:
We've got some scattered discussions on the issue of JSON-RPC  IDs. Let's move to matrix.